### PR TITLE
[anodized-fmt] test: basic fuzzing

### LINKED ISSUE #110: [anodized-fmt] test: prop-tests to increase coverage and reduce fixture complexity
Currently our testing strategy for `anodized-fmt` relies heavily on pre-determined test fixtures.

Extending our strategy with property-based testing could both increase coverage and reduce reliance on specific fixtures.

### PR DESCRIPTION
**Pull request by mkovaxx**
- Add test utils to fuzz `anodized-fmt`'s whitespace invariance.
- Set up a simple fuzz target: `fmt_ws_invariant`.
- Add a basic `fuzz/README.md`.

### CODE REVIEW

File: fuzz/fuzz_targets/fmt_ws_invariant.rs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        let config = Config::default();
+
+        let template = TEMPLATE.get().unwrap();
+        let default_input = template.generate(Variant::default());
+        let variant_input = template.generate(variant);
+
+        dbg!(&default_input);
+        dbg!(&variant_input);
+
+        let fmt_default = format_file(&default_input, &config).expect("formatting default");
```
Comment (mkovaxx) [inline fuzz/fuzz_targets/fmt_ws_invariant.rs:45]: @chammika Do you think it makes sense to call `format_file` here? Or maybe we could expose something like `format_spec_args(args: &str, config: &Config) -> Result<String>` to test spec formatting in isolation?

File: crates/test-util/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+[package]
```
Comment (mkovaxx) [inline crates/test-util/Cargo.toml:1]: Needed to create this non-published crate so that the fuzz targets can depend on it.

File: crates/test-util/src/fmt.rs [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+#[derive(Debug, Clone)]
+pub enum Span {
+    /// A fixed string.
+    F(String),
+    /// Zero or more whitespace characters.
+    Z,
+    /// One or more whitespace characters.
+    P,
+}
```
Comment (mkovaxx) [inline crates/test-util/src/fmt.rs:18]: @chammika I think this is a minimal approach we can start with.

The basic idea is this: formatting should be invariant to replacing a span of whitespace characters with more whitespace characters. So all we need is a template that can represent both verbatim character sequences (like code or comments) and markers at which whitespace should be inserted.

A minor complication is that we need to distinguish markers where the length of the whitespace span can be zero from markers where it must be at least one.

Here's an example:
```rust
let x = 42;
```

Omitting the space between `let` and `x` would make the fragment invalid, so it must be marked as "one or more", but all the other spaces could be omitted, so those need to be marked as "zero or more".

File: fuzz/fuzz_targets/fmt_whitespace.rs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+#[rustfmt::skip]
+fn make_template() -> Template {
+    Template::new()
+        .z().code("# [ spec (")
+        .z().line_comment(" the precondition")
+        .z().code("requires : x > 0 ,")
+        .z().line_comment(" the postcondition")
+        .z().code("ensures : * output > 0 ,")
+        .z().code(") ]")
+        .z().code("fn").p().code("func ( x : i32 ) -> i32 {")
```
Comment (mkovaxx) [inline fuzz/fuzz_targets/fmt_whitespace.rs]: Note the `.p()` in here between `fn` and `func`: that whitespace span must have at least one character.

I think we could make `.code()` a bit more intelligent so that this could be written simply as:
```rust
.z().code("fn func ( x : i32 ) -> i32 {")
```

**Comment by mkovaxx** [comment]
@chammika This is what I have so far, please try it out to see if it helps with developing `anodized-fmt`! :)

You should be able to run the new `fmt_whitespace` fuzz target like this:
```sh
cargo +nightly fuzz run fmt_whitespace
```

File: fuzz/fuzz_targets/fmt_whitespace.rs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+///         // postcondition
+///         ensures: *output > 0,
+///     )]
+///     fn func(x: i32) -> i32 { todo!() }
+///
+#[rustfmt::skip]
+fn make_template() -> Template {
+    Template::new()
+        .fixed("#[spec(")
+        //.line_comment(" precondition")
```
Comment (mkovaxx) [inline fuzz/fuzz_targets/fmt_whitespace.rs]: @chammika We'll be able to test comment formatting by enabling these.

File: fuzz/README.md [hunk 6]
```diff
@@ -0,0 +1,11 @@
+# Fuzz Targets for Testing Anodized Components
+
+## fmt_whitespace
+
+Tests that spec formatting is invariant to whitespace changes.
+
+**IMPORTANT**: Due to `proc-macro2` special-casing its behavior based on `cfg(fuzzing)` which `cargo-fuzz` sets by default, it must be run with the `--no-cfg-fuzzing` option as follows:
```
Comment (mkovaxx) [inline fuzz/README.md:7]: @chammika Heads up, this is pretty important! 😅

File: fuzz/fuzz_targets/fmt_ws_invariant.rs (hunk 1 above)
Comment (chammika) [reply fuzz/fuzz_targets/fmt_ws_invariant.rs:45]: `format_spec_args` only the macro is better right ? I will update

File: fuzz/README.md (hunk 6 above)
Comment (chammika) [reply fuzz/README.md:7]: It didn't take that long with fuzzing, isn't it ? :)
```bash
Diff < left / right > :
 #[spec(// precondition
< requires:x>0,// postcondition
>requires:x>0,// postcondition
 ensures:*output>0,)]
 fn func(x: i32) -> i32 { todo!() }
```
I will check this out...
Update: 
formatting skipped with `has_problematic_comments` for both `default_input` and `variant_input` but they are different due to the varient, NOT because they were formatted differently :)

Comment (chammika) [reply fuzz/README.md:7]: Got this spec fuzzed correctly in my PR #140
```rust
    Template::new()
        .fixed("#[spec(\n")
        .z().line_comment(" precondition")
        .z().code("requires : x > 0 ,").fixed("\n")
        .z().line_comment(" postcondition")
        .z().code("ensures : * output > 0 ,").fixed("\n")
        .z().fixed(")]\n")
        .fixed("fn func(x: i32) -> i32 { todo!() }\n")
```
Two things got fixed in above template
- `#spec(// comment` is skipped earlier with `has_problematic_comments`
- not possible to get inline comments, which are skipped as well

File: fuzz/fuzz_targets/fmt_whitespace.rs (hunk 5 above)
Comment (chammika) [reply fuzz/fuzz_targets/fmt_whitespace.rs]: This makes the template which was skipped (see below the template I came up with)

```txt
#[spec(// precondition
requires:...
```

We can make more templates when we support inline comments.

File: fuzz/README.md (hunk 6 above)
Comment (mkovaxx) [reply fuzz/README.md:7]: I've updated the template with your suggestions, thanks! :)

File: fuzz/fuzz_targets/fmt_ws_invariant.rs (hunk 1 above)
Comment (mkovaxx) [reply fuzz/fuzz_targets/fmt_ws_invariant.rs:45]: On a second thought, I think `format_file` is good. For example, it allows checking that the formatter does not alter code outside of specs.
